README.rst
This package provides a useful set of data generators that can be used in a
wide variety of applications, like testing and sample data generation.

Those generators are extensions to the ones provided in ``z3c.sampledata``.

See ``src/z3c/datagenerator/README.txt`` for usage documentation.


CHANGES.rst
=======
CHANGES
=======

2.1.2 (unreleased)
------------------

- Nothing changed yet.


2.1.1 (2014-04-07)
------------------

- Bug: FileDataGenerator used the full filename to generate the random seed.
  That is definitely going to break consistency/tests on someone else's system.


2.1.0 (2013-03-15)
------------------

- Switched code base to use ``random2``, so that Python 2 and 3 have the same
  output. No need for running different test files now.

- Simplified and unified test setup.


2.0.1 (2013-02-12)
------------------

- Updated manifest to include buildout and tox configuration files.

- Updated Trove classifiers.

- Renamed text files from *.txt to *.rst,s o Github renders them nicely.


2.0.0 (2013-02-06)
------------------

- Feature: Support for Python 3.

- Bug: Make sure that all files are closed properly.

- Bug: When generating a username, make sure it does not include any special
  characters passed through by the first or lastname, for example "O'Dorothy".

1.0.0 (2013-02-06)
------------------

- Feature: Added tests for all data generators.

- Feature: Added an ID data generator that can generate all sorts of IDs that
  could occur in systems.

- Feature: To properly support Windows, ``consistent_hash()`` returns an
  integer.

- Bug: The IPv4 generator ignored the seed making the generator "unstable".

0.0.3 (2008-12-03)
------------------

- Refined the seed generation further: zlib.crc32() in 32 bit Python can
  generate negative hashes, while 64 bit Python does not.  Enforced
  positive hashes.

- Began a test suite.


0.0.2 (2008-12-02)
------------------

- Use the crc32 function to hash random seeds so that the
  same random sequences are generated on both 32 bit and 64 bit
  builds of Python.


0.0.1 (2008-02-14)
------------------

- Initial Release


src/z3c/datagenerator/tests.py
##############################################################################
#
# Copyright (c) 2007 Zope Foundation and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
##############################################################################
"""Test Setup"""
import doctest
import re
import sys
import unittest
from zope.testing import renormalizing

checker = renormalizing.RENormalizing([
    # Python 3 unicode removed the "u".
    (re.compile("(?<![a-zA-Z])u('.*?')"),
     r"\1"),
    ])


def test_suite():
    return unittest.TestSuite((
        doctest.DocFileSuite(
                'README.rst',
                checker=checker,
                optionflags=doctest.NORMALIZE_WHITESPACE|doctest.ELLIPSIS,
                ),
        ))

if __name__ == '__main__':
    unittest.main(defaultTest='test_suite')


src/z3c/datagenerator/tests/test_doc.py
##############################################################################
#
# Copyright (c) 2007 Zope Foundation and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
##############################################################################
"""Test Setup"""
import doctest
import sys
import unittest
from zope.testing.doctestunit import DocFileSuite

def test_suite():
    filename = 'README3.rst' if sys.version_info[0] == 3 else 'README.rst'
    return unittest.TestSuite((
        DocFileSuite('../'+filename,
                     optionflags=doctest.NORMALIZE_WHITESPACE|doctest.ELLIPSIS,
                     ),
        ))

if __name__ == '__main__':
    unittest.main(defaultTest='test_suite')


src/z3c/datagenerator/net.py
##############################################################################
#
# Copyright (c) 2007 Zope Foundation and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
##############################################################################
"""Demographics Data Generators"""
__docformat__ = "reStructuredText"
import os
import random2 as random
import zope.interface

from z3c.datagenerator import demographics, generator, interfaces


@zope.interface.implementer(interfaces.IDataGenerator)
class IPv4DataGenerator(object):
    """IPv4 generator."""

    def __init__(self, seed):
        self.random = random.Random(generator.consistent_hash(seed+'ip'))

    def get(self):
        """Select a value from the values list and return it."""
        return '%.3i.%.3i.%.3i.%.3i' %(
            self.random.randint(1, 255), self.random.randint(0, 255),
            self.random.randint(0, 255), self.random.randint(0, 255))

    def getMany(self, number):
        """Select a set of values from the values list and return them."""
        return [self.get() for count in range(number)]


@zope.interface.implementer(interfaces.IDataGenerator)
class UsernameDataGenerator(object):
    """Username generator."""

    pattern = u'%(firstInitial)s%(lastName)s'

    def __init__(self, seed, pattern=None):
        self.random = random.Random(generator.consistent_hash(seed+'username'))
        self.firstNames = demographics.FirstNameGenerator(seed)
        self.lastNames = demographics.LastNameGenerator(seed)
        if pattern:
            self.pattern = pattern

    def get(self, firstName=None, lastName=None):
        """Select a value from the values list and return it."""
        fname = firstName or self.firstNames.get()
        lname = lastName or self.lastNames.get()
        return self.pattern % {
            'firstName':     ''.join(c for c in fname.lower() if c.isalnum()),
            'lastName':      ''.join(c for c in lname.lower() if c.isalnum()),
            'firstInitial':  fname[0].lower(),
            'lastInitial':   lname[0].lower(),
            'number':        self.random.randint(1, 100)}

    def getMany(self, number):
        """Select a set of values from the values list and return them."""
        return [self.get() for count in range(number)]


@zope.interface.implementer(interfaces.IDataGenerator)
class EMailDataGenerator(object):
    """E-Mail generator."""

    wordsFile = 'words.txt'
    tldsFile = 'gTLD.csv'

    pattern = '%(uname)s@%(domain)s%(tld)s'

    def __init__(self, seed):
        self.random = random.Random(generator.consistent_hash(seed+'enail'))
        self.usernames = UsernameDataGenerator(seed)
        self.words = generator.TextDataGenerator(seed, self.wordsFile)
        self.tlds = generator.CSVDataGenerator(seed, self.tldsFile)


    def get(self, username=None):
        """Select a value from the values list and return it."""
        return self.pattern %{
            'uname': username or self.usernames.get(),
            'domain': self.words.get(),
            'tld': self.tlds.get()[0]}


    def getMany(self, number):
        """Select a set of values from the values list and return them."""
        return [self.get() for count in range(number)]


src/z3c/datagenerator/interfaces.py
##############################################################################
#
# Copyright (c) 2007 Zope Foundation and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
##############################################################################
"""Data Generators Interfaces

$Id$
"""
__docformat__ = "reStructuredText"
import zope.interface
import zope.schema

class IDataGenerator(zope.interface.Interface):
    """Base functionality for data generators."""

    random = zope.interface.Attribute(
        '''An instance of the standard Python random number generator. This
        attribute is public, so that tests can implement predictable versions
        -- for example by setting the same seed all the time.''')

    def get(self):
        """Select a value from the values list and return it."""

    def getMany(self, number):
        """Select a set of values from the values list and return them."""


class IFileBasedGenerator(IDataGenerator):
    """Data generator using a single file extract data.

    Specific implementations include a simple line-based and a CSV one.
    """

    path = zope.interface.Attribute(
        'The path to the file providing the data.')


class IDateDataGenerator(IDataGenerator):
    """A date data generator.

    This generator creates dates/times between the start and end dates/times.
    """

    start = zope.schema.Datetime(
        title=u'Start Date/Time',
        description=u'This field descibres the earliest date/time generated.',
        required=True)

    end = zope.schema.Datetime(
        title=u'End Date/Time',
        description=u'This field descibres the latest date/time generated.',
        required=True)

    def get(self, start=None, end=None):
        """Create a new date between the start and end date.

        The start and end date/time can be overridden here, since you
        sometimes want to generate sequences of dates.
        """


src/z3c/datagenerator/demographics.py
##############################################################################
#
# Copyright (c) 2007 Zope Foundation and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
##############################################################################
"""Demographics Data Generators"""
__docformat__ = "reStructuredText"
import io
import os
import random2 as random
import zope.interface

from z3c.datagenerator import generator

class LastNameGenerator(generator.TextDataGenerator):
    """Last Name Generator"""

    def __init__(self, seed):
        super(LastNameGenerator, self).__init__(seed, 'lastnames.txt')


class FirstNameGenerator(generator.TextDataGenerator):
    """First Name Generator"""

    def __init__(self, seed):
        super(FirstNameGenerator, self).__init__(seed, 'firstnames.txt')


class SSNDataGenerator(object):
    """A social security data generator."""

    def __init__(self, seed):
        self.random = random.Random(generator.consistent_hash(seed+'ssn'))

    def get(self):
        """Compute a social security number."""
        randint = self.random.randint
        return u'%.3i-%.2i-%.4i' %(
            randint(1, 999), randint(1, 99), randint(1, 9999))

    def getMany(self, number):
        """Select a set of values from the values list and return them."""
        return [self.get() for count in range(number)]


class AddressDataGenerator(object):
    """An address data generator."""

    streetNamesFile = 'us-street-names.txt'
    streetPostfixFile = 'us-street-postfix.txt'
    citiesFile = 'us-cities.txt'
    statesFile = 'us-states.txt'
    apts = True

    def __init__(self, seed):
        self.random = random.Random(generator.consistent_hash(seed+'address'))
        path = os.path.dirname(__file__)

        with io.open(os.path.join(path, self.streetNamesFile), 'r',
                  encoding='latin-1') as file:
            self.streetNames = [e.strip() for e in file.readlines()]

        with io.open(os.path.join(path, self.streetPostfixFile), 'r',
                     encoding='latin-1') as file:
            self.streetPostfix = [e.strip() for e in file.readlines()]

        with io.open(os.path.join(path, self.citiesFile), 'r',
                     encoding='latin-1') as file:
            self.cities = [e.strip() for e in file.readlines()]

        with io.open(os.path.join(path, self.statesFile), 'r',
                     encoding='latin-1') as file:
            self.states = [e.strip() for e in file.readlines()]

    def getStreet(self):
        street = u'%i ' % self.random.randint(1, 2000)
        street += u'%s ' % self.random.sample(self.streetNames, 1)[0]
        street += self.random.sample(self.streetPostfix, 1)[0]
        if self.apts and self.random.random() < 0.3:
            street += u' Apt. %i' %self.random.randint(1, 30)
        return street

    def getCity(self):
        return self.random.sample(self.cities, 1)[0]

    def getState(self):
        return self.random.sample(self.states, 1)[0]

    def getZip(self):
        return u'%.5i' % self.random.randint(1000, 99999)

    def get(self):
        """Select a value from the values list and return it."""
        return self.getStreet(), self.getCity(), self.getState(), self.getZip()

    def getMany(self, number):
        """Select a set of values from the values list and return them."""
        return [self.get() for count in range(number)]


class PhoneDataGenerator(object):
    """A phone data generator."""

    template = u'%i-%.3i-%.4i'
    fivesAreaCode = False

    def __init__(self, seed, fivesAreaCode=False):
        self.random = random.Random(generator.consistent_hash(seed+'phone'))
        self.fivesAreaCode = fivesAreaCode

    def get(self):
        """Compute a social security number."""
        randint = self.random.randint
        areaCode = 555 if self.fivesAreaCode else randint(100, 999)
        return self.template %(areaCode, randint(1, 999), randint(1, 9999))

    def getMany(self, number):
        """Select a set of values from the values list and return them."""
        return [self.get() for count in range(number)]


src/z3c/datagenerator/generator.py
##############################################################################
#
# Copyright (c) 2007 Zope Foundation and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
##############################################################################
"""Data Generators"""
__docformat__ = "reStructuredText"
import csv
import datetime
import math
import io
import os
import random2 as random
import zope.interface
from zlib import crc32

from z3c.datagenerator import interfaces


def consistent_hash(buf):
    # Produce a hash of a string that behaves consistently in Python 32 and
    # 64 bit.  The "& 0xffffffff" interprets negative numbers as positive.
    return crc32(buf.encode('UTF-8')) & 0xffffffff


@zope.interface.implementer(interfaces.IDataGenerator)
class VocabularyDataGenerator(object):
    """Vocabulary-based data generator"""

    def __init__(self, seed, vocabulary):
        self.random = random.Random(consistent_hash(seed))
        self.vocabulary = vocabulary

    def get(self):
        """Select a value from the values list and return it."""
        return self.random.sample(self.vocabulary, 1)[0].value

    def getMany(self, number):
        """Select a set of values from the values list and return them."""
        return [term.value
                for term in self.random.sample(self.vocabulary, number)]


@zope.interface.implementer(interfaces.IFileBasedGenerator)
class FileDataGenerator(object):
    """Base functionality for a file data generator."""

    path = os.path.dirname(__file__)

    def __init__(self, seed, filename):
        justname = os.path.basename(filename)
        self.random = random.Random(consistent_hash(seed+justname))
        self.values = self._read(filename)

    def get(self):
        """Select a value from the values list and return it."""
        return self.random.sample(self.values, 1)[0]

    def getMany(self, number):
        """Select a set of values from the values list and return them."""
        return self.random.sample(self.values, number)


class CSVDataGenerator(FileDataGenerator):
    """CSV-based data generator."""

    def _read(self, filename):
        fullpath = os.path.join(self.path, filename)
        with io.open(fullpath, 'r', encoding='latin-1') as file:
            reader = csv.reader(file, delimiter=';')
            return [[cell for cell in row] for row in reader]


class TextDataGenerator(FileDataGenerator):
    """Text lines based data generator."""

    def _read(self, filename):
        fullpath = os.path.join(self.path, filename)
        with io.open(fullpath, 'r', encoding='latin-1') as file:
            return [e.strip() for e in file.readlines()]


@zope.interface.implementer(interfaces.IDateDataGenerator)
class DateDataGenerator(object):
    """A date data generator."""

    def __init__(self, seed, start=None, end=None):
        self.random = random.Random(consistent_hash(seed+'date'))
        self.start = start or datetime.date(2000, 1, 1)
        self.end = end or datetime.date(2007, 1, 1)

    def get(self, start=None, end=None):
        """Create a new date between the start and end date."""
        start = start or self.start
        end = end or self.end
        delta = end - start
        return start + datetime.timedelta(self.random.randint(0, delta.days))

    def getMany(self, number):
        """Select a set of values from the values list and return them."""
        return [self.get() for count in range(number)]


class IdDataGenerator(object):
    """An ID data generator."""

    prefix = 'ID'
    separator = '-'
    numbers = 4
    max_value = 99

    def __init__(self, seed, prefix='ID', separator='-', numbers=4,
                 max_value=99):
        self.random = random.Random(consistent_hash(seed + 'id'))
        self.prefix = prefix
        self.separator = separator
        self.numbers = numbers
        self.max_value = max_value
        self._num_format = '%.' + str(int(math.log10(self.max_value + 1))) + 'i'

    def get(self):
        """Compute a social security number."""
        randint = self.random.randint
        value = self.prefix
        value += self.separator.join(
            self._num_format % randint(0, self.max_value)
            for idx in range(self.numbers))
        return value

    def getMany(self, number):
        """Select a set of values from the values list and return them."""
        return [self.get() for count in range(number)]
